src/main/java/org/spongepowered/api/text/template/TextTemplate.java

```diff
@@ -0,0 +1,244 @@
+     * takes any type as it's input and returns a {@link TextElement} as
+     * it's output.
+     *
+     * @param name Name of parameter
+     * @param type Type of parameter
+     * @param f Function that takes "type" as the input and returns Text as the output
+     * @param optional Whether this argument is allowed to be omitted
+     * @param <T> Input type of argument
+     * @return Newly constructed Arg
+     */
+    public static <T> Arg<T> arg(String name, Class<T> type, Function<T, TextElement> f, boolean optional) {
+        checkNotNull(name, "name");
+        checkArgument(!name.isEmpty(), "name");
+        checkNotNull(type, "argument type");
+        checkNotNull(f, "function");
+        return new Arg<>(name, type, f, optional);
+    }
+
+    /**
+     * Constructs a new {@link Arg} to be supplied to {@link #of(Object...)}.
+     * The argument evaluates it's output by applying a {@link Function} that
+     * takes any type as it's input and returns a {@link TextElement} as it's output.
+     *
+     * @param name Name of parameter
+     * @param type Type of parameter
+     * @param f Function that takes "type" as the input and returns Text as the output
+     * @param <T> Input type of argument
+     * @return Newly constructed Arg
+     */
+    public static <T> Arg<T> arg(String name, Class<T> type, Function<T, TextElement> f) {
+        return arg(name, type, f, false);
+    }
+
+    /**
+     * Constructs a new {@link Arg} to be supplied to {@link #of(Object...)}.
+     * This argument expects a {@link TextElement} parameter.
+     *
+     * @param name Name of parameter
+     * @param optional Whether this argument is allowed to be omitted.
+     * @return Newly constructed Arg
+     */
+    public static Arg<TextElement> arg(String name, boolean optional) {
+        return arg(name, TextElement.class, t -> t, optional);
+    }
+
+    /**
+     * Constructs a new {@link Arg} to be supplied to {@link #of(Object...)}.
+     * This argument expects a {@link TextElement} parameter.
+     *
+     * @param name Name of parameter
+     * @return Newly constructed Arg
+     */
+    public static Arg<TextElement> arg(String name) {
+        return arg(name, false);
+    }
+
+    /**
+     * Represents a variable element within a TextTemplate. Each argument has
+     * an input type and all arguments return a {@link TextElement} as it's
+     * output.
+     *
+     * @param <T> Input type
+     */
+    static class Arg<T> {
+        final String name;
+        final Class<T> type;
+        final Function<T, TextElement> f;
+        final boolean optional;
+
+        /**
+         * Evaluates the argument for the specified parameter.
+         *
+         * @param param Parameter to evaluate
+         * @return TextElement output
+         */
+        @SuppressWarnings("unchecked")
+        TextElement eval(Object param) {
+            Class<?> paramType = param.getClass();
```

Comment:
Not really, this method is not accessible externally and is only called with non-null params. Null parameters are skipped if they are optional.